Experiment with angular-cache

oliverguenther opened the pull request:
This PR introduced a memory/sessionStorage cache using `angular-cache`. While angular has its own `$cacheFactory`, its options are rather limited and provide no integration with, e.g., the local and sessionStorage.

One prominent example where we should employ frontend-side caching
is the fetching of associated users in activities, which is the main driver for this PR.

For a large number of comments of the same user, that user is retrieved
once per activity, which causes a drastic overhead.

This commit introduces a simple wrapper around angular-cache to store
values in the sessionStorage.
### TODOS / Open questions
- [x] Multiple cache stores (e.g., one for persistence / seldomly-changed values in localStorage)
- [x] ~~Storing the promise in sessionStorage feels like magic, as I think some browsers limit the storage to string only. We should keep these things in memory, perhaps.~~ I'm now storing the returned data in the sessionStorage.

Note: Opening a page in a new tab or window will result in a fresh sessionStorage, so this is rather a `tab / window storage`.

oliverguenther committed:
Experiment with angular-cache

One prominent example where we can employ frontend-side caching
is the fetching of associated users in activities.

For a large number of comments of the same user, that user is retrieved
once per activity, which causes a drastic overhead.

This commit introduces a simple wrapper around angular-cache to store
values in the sessionStorage.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-angular-cache.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-angular-cache.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
From an code standpoint (and this explicitly excludes framework knowledge), this looks like a rather nice solution. The concerns seem to be separated quite nicely, I could only discuss some naming decisions, but that is always possible :horse: 

Still this is the only thing I can comment on here, someone closer to JavaScript and Angular will probably see errors/problems that I would miss. But to me, this looks like the right direction. :horse:

NobodysNightmare commented on frontend/app/services/cache-service.js:
```diff
@@ -0,0 +1,77 @@
+  var cacheName = 'openproject-cache';
+  var _cache = CacheFactory(cacheName, {
+    maxAge: 30 * 60 * 1000, // 30 mins
+    storageMode: 'sessionStorage'
```
After reading a little more stuff up and getting other stuff out of my memory:

I would have expected either `memory` or `localStorage` here:

Memory is the least persistent one, a page refresh will clear it. This is the least invasive option, because good old F5 will get me to a well-known state.

If we want to cache across refresh-cycles we would need one of the other storage options. I (personally) did not know the difference between sessionStorage and localStorage, just knowing the latter one. If I understand the difference correctly, the localStorage will work browser wide, ~~while the session storage will work "per window"~~. If that is the only difference, I would suggest using the localStorage rather than the session storage, since this will also cache across multiple tabs.

edit: apparently the session cache lives as long as you keep your browser open, which is a much more useful limitation. Now I also see why you chose session rather than local storage.

Still my favorite for now is the memory based storage, since it is the least invasive one and it still works well if our application is used one-page-style.

oliverguenther committed:
Temporarily store promises until they are resolved

The original commit simply stored promises into the cache,
which seemed to work in browsers but failed with karma/PhantomJS.

I'm not yet certain why that is, but I wanted to get rid of complex
objects in sessionStorage anyway, so this commit does:

* temporarily store running promises for a given key
* cache the resolved data directly

NobodysNightmare commented:
Core specs fail with the same single error. Looks kind of reasonable...

basiszwo commented:
@oliverguenther what about the open task? is it about adding support for different cache stores in general?

oliverguenther commented:
There is a legitimate spec failure due to the way the cache operates. It simply takes the API resource location (i.e., `User ID#1` => `api/v3/users/1`) as the cache key. As FactoryGirl will create sequences of the same user **with the same ID**, they are cached by the Frontend.

We will either need to:
1. Disable the cache for tests
2. Clear the cache explicitly where necessary.
3. Ensure increasing IDs (unfeasible)

However, as this is a side project in my free time, I did not yet have the time to discuss this with either you or @ulferts, thus the failure is not yet addressed.

basiszwo commented:
I would just disable this caching in dev and test mode. 
This is what rails does too.

## 

On 29 Sep 2015, at 14:49, Oliver Günther wrote:

> There is a legitimate spec failure due to the way the cache operates. It simply takes the API resource location (i.e., `User ID#1` => `api/v3/users/1`) as the cache key. As FactoryGirl will create sequences of the same user **with the same ID**, they are cached by the Frontend.
> 
> We will either need to:
> 1. Disable the cache for tests
> 2. Clear the cache explicitly where necessary.
> 3. Ensure increasing IDs (unfeasible)
> 
> However, as this is a side project in my free time, I did not yet have the time to discuss this with either you or @ulferts, thus the failure is not yet addressed.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/opf/openproject/pull/3553#issuecomment-144048389

oliverguenther committed:
Disable CacheService for testing

oliverguenther commented:
I've disabled it for test for the time being in order to avoid a different behavior on production and dev and to aid discover frontend caching issues early on.

Tests should now pass, and this is ready to merge from my side.

oliverguenther committed:
Allow more than one cache

oliverguenther commented:
@furinvader If you can spare the time, can you have a look whether this is something we should integrate?

The main motivation of this PR is to prevent the many subsequent requests to the same `/api/v3/users/:id` resource in the activity stream of a WP:
![bildschirmfoto 2015-10-26 um 07 41 54](https://cloud.githubusercontent.com/assets/459462/10722853/1a0d422e-7bb5-11e5-84cd-723ab0373871.png)

oliverguenther committed:
Globally disable CachService in run block

Avoid checking the state of the cache in every resource loading,
but rather disable it upon startup.

oliverguenther commented:
After a discussion with @furinvader , I have extracted disabling the CacheService into the run block of `openproject-app.js` to disable the cache globally and avoid the window depedency in the service.
